sway-lib-std/src/assert.sw

```diff
@@ -10,3 +10,12 @@ pub fn assert(a: bool) {
         ()
     }
 }
+
+/// wrapper for `assert` that allows passing a custom revert value `v` if condition `c` is not true.
+pub fn require(c: bool, v: u64) {
```

Comment:
Make this generic, and use `LOGD` followed by a revert. You can use `is_reference_type` and `size_of` to know how many bytes to log. For simplicity, just use `LOGD` all the time even if it's a word.

Reply:
What would get logged exactly?  This function looks to me like an `assert()` but instead of `revert(0)` it's `revert(x)` which makes tracking down which assert failed a bit easier.

You're suggesting a third argument to add some verbosity?  Or to remove `v` and log something else and still `revert(0)`?

Reply:
Similar to Solidity custom errors (https://blog.soliditylang.org/2021/04/21/custom-errors/), `require`'s second parameter should be a struct which can eventually be part of the JSON ABI. In the meantime, at least it can be logged even though the logged value can't be decoded by the SDK.

Reply:
hmm. If we make `require` generic, I think we need to also be able to add the constraint that the generic value either is a `u64` or is a struct/enum with a field/variant with a `u64` type which can be passed to the inner call to `revert`.

Reply:
Perhaps something along these lines:
`require`'s second param is a struct type (ie: `CustomError` provided by the `assert` module in the stdlib. Users can instantiate as many CustomErrors as they need in their contract, and there will always be a  `u64` value available to pass to `revert`:
```
pub struct CustomError<T> {
    Id: ContractId, // use to log the current contract id 
    RevertValue: u64, // value to pass to `revert`
    AdditionalData: T // user can log out generic data to keep this flexible
}
    
    
```

Reply:
There's no need for a `u64` value, just set that to a hard-coded standard value for indicating a custom error. See Solidity panic codes as an example: https://docs.soliditylang.org/en/v0.8.4/control-structures.html#panic-via-assert-and-error-via-require

`LOGD` already logs the contract ID in the receipt: https://github.com/FuelLabs/fuel-specs/blob/84cff9a590a28d7c5e6cb62d91b52e8e30c6ef67/specs/vm/opcodes.md#logd-log-data-event

Reply:
This can be as simple as `fn require<T>(c: bool, v: T)`, then just `LOGD` `v`, followed by a `RVRT` with some error code that's documented to mean "look at the previous logd receipt".